registration/include/pcl/registration/impl/gicp.hpp

```diff
@@ -445,87 +521,55 @@ GeneralizedIterativeClosestPoint<PointSource, PointTarget>::computeTransformatio
+      // M = temp^-1
+      M = temp.inverse();
     }
-    // Resize to the actual number of valid correspondences
-    source_indices.resize(cnt);
-    target_indices.resize(cnt);
+
     /* optimize transformation using the current assignment and Mahalanobis metrics*/
     previous_transformation_ = transformation_;
     // optimization right here
     try {
       rigid_transformation_estimation_(
           output, source_indices, *target_, target_indices, transformation_);
-      /* compute the delta from this iteration */
-      delta = 0.;
-      for (int k = 0; k < 4; k++) {
-        for (int l = 0; l < 4; l++) {
-          double ratio = 1;
-          if (k < 3 && l < 3) // rotation part of the transform
-            ratio = 1. / rotation_epsilon_;
-          else
-            ratio = 1. / transformation_epsilon_;
-          double c_delta =
-              ratio * std::abs(previous_transformation_(k, l) - transformation_(k, l));
-          if (c_delta > delta)
-            delta = c_delta;
-        }
-      }
     } catch (PCLException& e) {
       PCL_DEBUG("[pcl::%s::computeTransformation] Optimization issue %s\n",
                 getClassName().c_str(),
                 e.what());
       break;
     }
+
+    // Transform the data
+    pcl::transformPointCloud(output, *input_transformed, transformation_);
+
     nr_iterations_++;
 
     if (update_visualizer_ != 0) {
-      PointCloudSourcePtr input_transformed(new PointCloudSource);
-      pcl::transformPointCloud(output, *input_transformed, transformation_);
       update_visualizer_(*input_transformed, source_indices, *target_, target_indices);
     }
-
-    // Check for convergence
-    if (nr_iterations_ >= max_iterations_ || delta < 1) {
-      converged_ = true;
-      PCL_DEBUG("[pcl::%s::computeTransformation] Convergence reached. Number of "
-                "iterations: %d out of %d. Transformation difference: %f\n",
-                getClassName().c_str(),
-                nr_iterations_,
-                max_iterations_,
-                (transformation_ - previous_transformation_).array().abs().sum());
-      previous_transformation_ = transformation_;
-    }
-    else
-      PCL_DEBUG("[pcl::%s::computeTransformation] Convergence failed\n",
+    converged_ = static_cast<bool>((*convergence_criteria_));
+  } while (convergence_criteria_->getConvergenceState() ==
+           pcl::registration::GICPConvergenceCriteria<
+               float>::CONVERGENCE_CRITERIA_NOT_CONVERGED);
+  if(converged_) {
+    PCL_DEBUG("[pcl::%s::computeTransformation] Convergence reached. Number of "
+              "iterations: %d out of %d. Transformation difference: %f\n",
+              getClassName().c_str(),
+              nr_iterations_,
+              max_iterations_,
+              (transformation_ - previous_transformation_).array().abs().sum());
+    previous_transformation_ = transformation_;
+  }
+  else {
+    PCL_DEBUG("[pcl::%s::computeTransformation] Convergence failed\n",
```

Comment:
Originally it shows `Convergence failed` in each iteration, now it shows `Convergence failed` after breaking the while loop